src/history/test/HistoryTests.cpp

```diff
@@ -713,6 +713,152 @@ TEST_CASE("History catchup with different modes",
+    auto& lm = app->getLedgerManager();
+
+    auto& hm = app->getHistoryManager();
+    auto& herder = static_cast<HerderImpl&>(app->getHerder());
+
+    auto runCatchup = [&](uint32_t expectedDestination) {
+        auto startCatchupMetrics = catchupSimulation.getCatchupMetrics(app);
+
+        auto expectedCatchupWork =
+            catchupSimulation.computeCatchupPerformedWork(
+                lm.getLastClosedLedgerNum(),
+                CatchupConfiguration{expectedDestination,
+                                     app->getConfig().CATCHUP_RECENT,
+                                     CatchupConfiguration::Mode::ONLINE},
+                *app);
+
+        catchupSimulation.crankUntil(
+            app, [&]() { return app->getCatchupManager().catchupWorkIsDone(); },
+            std::chrono::seconds{
+                std::max<int64>(expectedCatchupWork.mTxSetsApplied + 15, 60)});
+
+        auto endCatchupMetrics = catchupSimulation.getCatchupMetrics(app);
+        auto catchupPerformedWork =
+            CatchupPerformedWork{endCatchupMetrics - startCatchupMetrics};
+
+        REQUIRE(catchupPerformedWork == expectedCatchupWork);
+    };
+
+    // Externalize (to the catchup LM) the range of ledgers between initLedger
+    // and as near as we can get to the first ledger of the block after
+    // initLedger (inclusive), so that there's something to knit-up with. Do not
+    // externalize anything we haven't yet published, of course.
+    uint32_t firstLedgerInCheckpoint =
+        hm.firstLedgerAfterCheckpointContaining(initLedger);
+
+    uint32_t triggerLedger = hm.ledgerToTriggerCatchup(firstLedgerInCheckpoint);
+
+    SECTION("A non-checkpoint ledger re-triggers CatchupWork")
+    {
+        // The app hears initLedger, ..., triggerLedger + bufferLedgers - 1.
+        // This is enough to kick out some ledgers from mSyncingLedgers.
+        for (uint32_t n = initLedger; n <= triggerLedger + bufferLedgers - 1;
+             ++n)
+        {
+            catchupSimulation.externalizeLedger(herder, n);
+        }
+
+        // Let the catchup run.
+        // We expect that it'll land at the checkpoint containing (initLedger -
+        // 1) without applying the buffered ledgers. This is because
+        // mSyncingLedgers must have popped some elements in order to prevent it
+        // from growing exponentially, and thus there is a gap between the LCL
+        // and mSyncingLedgers.
+        runCatchup(hm.checkpointContainingLedger(initLedger - 1));
+
+        // In order to close the gap between the LCL and mSyncingLedgers,
+        // we need to run a CatchupWork again.
+        // processLedger is responsible for that, and thus we call it
+        // by externalizing another ledger.
+        catchupSimulation.externalizeLedger(herder,
+                                            triggerLedger + bufferLedgers);
+        runCatchup(
+            hm.checkpointContainingLedger(initLedger + bufferLedgers - 2));
```

Comment:
Why is `initLedger + bufferLedgers - 2` the expected destination? I expected it to be `triggerLedger + bufferLedgers` since you get to apply all the remaining ledgers now and get in sync